server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -3286,19 +3321,101 @@ private void doCheckIndex() throws IOException {
     }
 
     Engine getEngine() {
-        Engine engine = getEngineOrNull();
-        if (engine == null) {
-            throw new AlreadyClosedException("engine is closed");
+        engineLock.readLock().lock();
+        try {
+            return getCurrentEngine(false);
+        } finally {
+            engineLock.readLock().unlock();
         }
-        return engine;
     }
 
     /**
      * NOTE: returns null if engine is not yet started (e.g. recovery phase 1, copying over index files, is still running), or if engine is
      * closed.
      */
     public Engine getEngineOrNull() {
-        return this.currentEngineReference.get();
+        engineLock.readLock().lock();
+        try {
+            return getCurrentEngine(true);
+        } finally {
+            engineLock.readLock().unlock();
+        }
+    }
+
+    public <R> R withMutableEngine(Function<Engine, R> operation) {
+        return withEngine(operation, true, false);
+    }
+
+    public <R> R withMutableEngineOrNull(Function<Engine, R> operation) {
+        return withEngine(operation, true, true);
+    }
+
+    public <R> R withImmutableEngine(Function<Engine, R> operation) {
+        return withEngine(operation, false, false);
+    }
+
+    /**
+     * Executes an operation while preventing the shard's engine instance to be changed or closed during the execution. The parameter
+     * {@code requiredMutability} can be used to force the engine to be reset to a given mutable or immutable state before executing the
+     * operation. The parameter {@code allowNoEngine} is used to allow the operation to be executed with a null engine instance, in which
+     * case the {@code requiredMutability} is ignored. When {@code allowNoEngine} is set to {@code `false`} the method will throw an
+     * {@link AlreadyClosedException} if the current engine is null and won't reset the engine to the required mutability state.
+     *
+     * @param operation
+     * @param requiredMutability
+     * @param allowNoEngine
+     * @return
+     * @param <R>
+     */
+    private <R> R withEngine(Function<Engine, R> operation, boolean requiredMutability, boolean allowNoEngine) {
+        assert ClusterApplierService.assertNotClusterStateUpdateThread("IndexShard.withEngine() can block");
+        assert MasterService.assertNotMasterUpdateThread("IndexShard.withEngine() can block");
+        assert Transports.assertNotTransportThread("IndexShard.withEngine() can block");
+        assert operation != null;
+
+        engineLock.readLock().lock();
+        var release = true;
+        try {
+            var engine = getCurrentEngine(allowNoEngine);
+            if (engine != null && (engine.isMutable() == requiredMutability) == false) {
```

Comment:
I wonder if there is a way where we can instead check something like `engine.resetPending()`, i.e., care less about the specific case and more about reset in progress?